### ISSUE TITLE
Hardlink to symlink behaviour differs from Linux

### ISSUE THREAD

**Issue by vizanto** [joyent/illumos-joyent#84]
Creating a hardlink to a symlink is allowed on linux.
Creating a hardlink to a symlink which points to a directory also, but not on illumos.

##### On illumos

```sh
[root@nixos:/nix/store/whut]# ln -s foo bar

[root@nixos:/nix/store/whut]# ln bar bar2

[root@nixos:/nix/store/whut]# ls -Al
lrwxrwxrwx 2 root root 3 Nov 16 17:28 bar -> foo
lrwxrwxrwx 2 root root 3 Nov 16 17:28 bar2 -> foo


[root@nixos:/nix/store/whut]# mkdir foo


[root@nixos:/nix/store/whut]# ls -Al
total 2
lrwxrwxrwx 2 root root 3 Nov 16 17:28 bar -> foo
lrwxrwxrwx 2 root root 3 Nov 16 17:28 bar2 -> foo
drwxr-xr-x 2 root root 2 Nov 16 17:29 foo


# Wait for it...!

[root@nixos:/nix/store/whut]# ln bar bar3
ln: failed to create hard link ‘bar3’ => ‘bar’: Operation not permitted

# Zing!
```

##### On linux:

```sh
[root@elon:/whut]# ln -s foo bar

[root@elon:/whut]# ln bar bar2

[root@elon:/whut]# ls -Al
total 1
lrwxrwxrwx 2 root root 3 Nov 16 18:37 bar -> foo
lrwxrwxrwx 2 root root 3 Nov 16 18:37 bar2 -> foo

[root@elon:/whut]# mkdir foo

[root@elon:/whut]# ls -Al
total 13
lrwxrwxrwx 2 root root 3 Nov 16 18:37 bar -> foo
lrwxrwxrwx 2 root root 3 Nov 16 18:37 bar2 -> foo
drwxr-xr-x 2 root root 2 Nov 16 18:38 foo

[root@elon:/whut]# ln bar bar3

[root@elon:/whut]# ls -Al
total 14
lrwxrwxrwx 3 root root 3 Nov 16 18:37 bar -> foo
lrwxrwxrwx 3 root root 3 Nov 16 18:37 bar2 -> foo
lrwxrwxrwx 3 root root 3 Nov 16 18:37 bar3 -> foo
drwxr-xr-x 2 root root 2 Nov 16 18:38 foo
```

Directly hardlinking a directory is not allowed btw:
```
[root@elon:/whut]# ln foo baz
ln: ‘foo’: hard link not allowed for directory
```

#### Background:
The nix package manager has a `nix-store --optimize` operation that deduplicates files using hardlinks.  On linux this program deduplicates symlinks as well.

**Comment by vizanto** [comment]
My NixOS lx-zone release is kind of waiting for this... :)

**Comment by pfmooney** [comment]
Open as [OS-5094](https://smartos.org/bugview/OS-5094)
I've drafted a fix which should be reviewed and put back later today.

**Comment by pfmooney** [comment]
This should be fixed in joyent/illumos-joyent@8843c35f721fa2818edf0a3ca6047c03d7021f30

**Comment by vizanto** [comment]
Looking forward to the next release, thanks!
